doc/SAI-Proposal-IPv6_Segment_Routing-1.md

```diff
@@ -0,0 +1,207 @@
+-------------|-----------------------------------------------------------------
+ Authors     | Cavium Inc.
+ Status      | In review
+ Type        | Standards track
+ Created     | 04/14/2017
+ Updated     | 05/11/2017
+ SAI-Version | 1.2.0
+
+-------------------------------------------------------------------------------
+
+## Overview ##
+
+Segment Routing (SR) allows a node to steer a packet through a controlled set of instructions, called segments, by prepending an SR header to the packet.  A segment can represent any instruction, topological or service-based.  SR allows to enforce a flow through any path (topological, or application/service based) while maintaining per-flow state only at the ingress node to the SR domain.
+
+This document covers IPv6-based Segment Routing as per IETF Drafts:
+1. https://tools.ietf.org/html/draft-ietf-6man-segment-routing-header-06.  
+2. https://tools.ietf.org/html/draft-filsfils-spring-srv6-network-programming-00
+3. https://tools.ietf.org/html/draft-filsfils-spring-segment-routing-policy-00
+
+This specification proposes the following points:
+1. Introduce the concept of Segment Routing using IPv6.
+2. Introduce behavioral model modifications to support SR source, transit, and endpoint.
+3. Introduce SAI APIs to define SR properties including segment and TLV definitions.
+
+## Behavioral Model
+
+In order to add IPv6 Segment Routing, it requires three mechanisms:
+1. Way to specify which flows will be marked for SR source or transit
+2. Way to add SR header (source / transit) or modify / remove SR header (endpoint) before normal L2/L3 processing
+
+For the first mechanism, ACL is used to match on any flow characteristics to determine the policy "color".  The color and packet DIP is used in the ACL match lookup to determine the Policy's Binding Segment ID (BSID) which is enabled in the next-hop group table. Alternatively, one can use the route table to just use DIP alone to determine the BSID.  The BSID is used in the next-hop group table to group Segment ID Lists (SID Lists) and ECMP hash between the possible paths
+
+For the second mechanism, the route / next-hop group / next-hop tables are used to identify the SR DIP and manipulate via endpoint function or source/transit SID lists as objects in the next-hop table.  Afterwards the flow will head to the egress pipeline for futher lookup on the modified packet.  Thus, the only structural changes needed is in the next-hop table to handle the transit, source, and endpoint actions.
+
+The below figures shows the additional logic in the Ingress ACL and Route lookup in the behavioral pipeline to support this.
+
+![SAI v6SR bm](figures/sai_v6SR_bm.png "Figure 1: Behavioral Model Addition. ")
+__Figure 1: Behavioral Model Addition.__
+
+![SAI v6SR bm1](figures/sai_v6SR_bm1.png "Figure 2: Source Behavior ")
+__Figure 2: Source Behavior.__
+
+![SAI v6SR bm2](figures/sai_v6SR_bm2.png "Figure 3: Transit Behavior ")
+__Figure 3: Transit Behavior.__
+
```

Comment:
As you know, there's no 'transit' specialized behavior per-se in regular SRv6: you're either a 'source' or an 'endpoint'. Real 'transit' nodes are those that handle an IPv6 packet with a DA/DIP different than themselves, so they are not expected to process the SRH, thus, they are not existing from a SRv6 standpoint. All the other nodes, excepting the 'source', are SRv6 'endpoints'. But i guess that your 'transit' semantic is different than mine in the sense that you are referring to nodes that are neither 'ingress' nor 'egress' from a traffic-flow perspective. I got it, no issues there. The reason I'm making this point is because in your 'transit' diagram, you are drawing a "set sidlist" action as part of the nh-table processing, which can be understood as if real srv6-transit nodes require a new sid-list to be provided by the user to allow the srh-pop operation. Please confirm that all 'srv6-transit' nodes are simply expected to execute the following instructions (further below) as per SRv6 drafts, and that there's no need to pass a new/shorter sid-list to SAI. If my understanding is correct, please adjust/remove the transit diagram if that's ok with you.

   1. IF SegmentsLeft > 0 THEN
   2.    decrement SL
   3.    update the IPv6 DA with SRH[SL]
   4.    FIB lookup on updated DA
   5.    forward accordingly to the matched entry
   6.  ELSE
   7.    drop the packet

[ After writing all the above, i now realize that by 'transit' you might be referring to the semantics exposed in SR-TE/policy draft --to allow insertions of sid-lists within an incoming sid-list. If that's the case, please ignore my comment above, and write a note to prevent misunderstandings like the one i may have incurred ;-) ]